src/Paprika/Data/SlottedArray.cs

```diff
@@ -904,80 +908,199 @@ private void MarkAsDeleted(int index)
+        var jump = DoubleVectorSize / sizeof(ushort);
+        var aligned = AlignToDoubleVectorSize(_header.Low) / sizeof(ushort);
+        ref var d = ref Unsafe.As<byte, ushort>(ref MemoryMarshal.GetReference(_data));
 
-        var i = 0;
+        // The data will be moved from index [start..end] to [writeAt..] where writeAt < start <= end.
+        var startIndex = NotFound;
+        var endIndex = NotFound;
+        var writeAtIndex = NotFound;
 
-        // Search for the first deleted
-        for (; i < count; i++)
+        for (var i = HashesPerVector; i < aligned; i += jump)
         {
-            var slot = GetSlotRef(i);
-            if (slot.IsDeleted)
-            {
-                writeAt = i; // write at this slot
-                readTo = slot.ItemAddress; // mark it as read already
-                writtenTo = (ushort)(i == 0 ? _data.Length : GetSlotRef(i - 1).ItemAddress); // memoize the previous
+            var deletedMask = CreateDeletedBitmask(ref d, i);
 
-                // move to next
-                i++;
-                break;
+            if (deletedMask != 0 && i + jump >= aligned)
+            {
+                // Clear the bits beyond the last valid slot.
+                var alignedCount = aligned / VectorsByBatch;
+                var toClear = alignedCount - count;
+                var mask = (1U << SlotsPerVector - toClear) - 1;
+                deletedMask &= mask;
             }
 
-            alive++;
-        }
+            while (deletedMask != 0)
+            {
+                // Consume the next set bit.
+                var setBitIndex = BitOperations.TrailingZeroCount(deletedMask);
+                deletedMask ^= 1U << setBitIndex;
+                alive--;
 
-        for (; i < count; i++)
-        {
-            var slot = GetSlotRef(i);
-            var addr = slot.ItemAddress;
+                if (startIndex == NotFound)
+                {
+                    startIndex = setBitIndex + (i - HashesPerVector) / VectorsByBatch + 1;
 
-            if (!slot.IsDeleted)
-            {
-                alive++;
+                    if (writeAtIndex == NotFound)
+                    {
+                        writeAtIndex = startIndex - 1;
+                    }
+
+                    if (deletedMask == 0)
+                    {
+                        // No more deleted slots, move on to the next vector to find the end.
+                        break;
+                    }
 
-                var length = readTo - addr;
+                    setBitIndex = BitOperations.TrailingZeroCount(deletedMask);
+                    deletedMask ^= 1U << setBitIndex;
+                    alive--;
+                }
 
-                if (length > 0)
+                endIndex = setBitIndex + (i - HashesPerVector) / VectorsByBatch - 1;
+
+                // In case of consecutive deleted slots, find the next non-consecutive set bit.
+                while (startIndex == endIndex + 1 && deletedMask != 0)
```

Comment:
Why this requires a loop? Can't it be that we find it bitwise without the loop?

Reply:
Nice catch! :grinning:  
Gave more thought and converted the loop into bit operations.